### ISSUE TITLE
[gatsby-image] Art-directing multiple images gives duplicate key warning

### ISSUE THREAD

**Issue by dante-blitz** [gatsbyjs/gatsby#16762]
## Description

I have followed the example for showing different images at different breakpoints. All works as expected but I'm getting the following warning in the console: 

Encountered two children with the same key, `https://www.mydomain.com/image.jpg?auto=compress%2Cformat&fm=jpg`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.

### Steps to reproduce

```
  const sources = [
    post.mobileImage.fluid,
    {
      ...post.xlImage.fluid,
      media: `(min-width: ${theme.breakpoints.xl})`,
    },
  ]
```

```
<Img fluid={sources} />
```

### Expected result

No same key warnings.

### Actual result

Same key warnings.

### Environment

  System:
    OS: Windows 10
    CPU: (8) x64 Intel(R) Xeon(R) CPU E5-1620 0 @ 3.60GHz
  Binaries:
    Yarn: 1.16.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.4.1 - C:\Program Files\nodejs\npm.CMD
  Languages:
    Python: 2.7.15 - /c/Users/user/.windows-build-tools/python27/python
  Browsers:
    Edge: 44.18362.267.0
  npmPackages:
    gatsby: ^2.13.70 => 2.13.70 
    gatsby-image: ^2.2.9 => 2.2.9 
    gatsby-plugin-eslint: ^2.0.5 => 2.0.5 
    gatsby-plugin-material-ui: ^2.1.5 => 2.1.5 
    gatsby-plugin-offline: ^2.2.6 => 2.2.6 
    gatsby-plugin-react-helmet: ^3.1.3 => 3.1.3 
    gatsby-plugin-react-svg: ^2.1.2 => 2.1.2 
    gatsby-plugin-sharp: ^2.2.12 => 2.2.12 
    gatsby-plugin-sitemap: ^2.2.6 => 2.2.6 
    gatsby-plugin-styled-components: ^3.1.2 => 3.1.2 
    gatsby-plugin-webpack-size: ^1.0.0 => 1.0.0 
    gatsby-source-datocms: 2.1.15 => 2.1.15 
    gatsby-transformer-sharp: ^2.2.6 => 2.2.6

**Comment by kimbaudi** [comment]
@dante-blitz - I think I know why you are receiving this duplicate key warning. It's kind of hard for me to explain, but basically the `maxWidth` that you specify in your graphql query for the image is larger than the actual image causing the image to be duplicated. Let me try to clarify with an example.

Suppose I have an image called `foo.png` with dimension 800px x 600px.

Now, I suppose my graphql query looks like the following:

```graphql
  query {
    mobileImage: file(relativePath: { eq: "foo.png" }) {
      childImageSharp {
        fluid(maxWidth: 800, quality: 100) {
          ...GatsbyImageSharpFluid
        }
      }
    }
    desktopImage: file(
      relativePath: { eq: "foo.png" }
    ) {
      childImageSharp {
        fluid(maxWidth: 1600, quality: 100) {
          ...GatsbyImageSharpFluid
        }
      }
    }
  }
```

This is going to cause the duplicate key warning because the image "foo.png" can only be rendered with a max width of 800px. So when you specify `fluid(maxWidth: 1600`, that is also going to render another image with a max width of only 800px, which is going to cause the "duplicate key" warning.

So, how do you get rid of the warning? Make sure your graphql query does not cause duplicate images to be rendered by either removing the query with `maxWidth: 1600` or making sure the `maxWidth`s set will produce images of different sizes:

a possible solution (remove `desktopImage`):

```graphql
  query {
    mobileImage: file(relativePath: { eq: "foo.png" }) {
      childImageSharp {
        fluid(maxWidth: 800, quality: 100) {
          ...GatsbyImageSharpFluid
        }
      }
    }
  }
```

another possible solution - make sure images rendered will have unique dimensions (i.e., 600px and 800px are unique image sizes):

```graphql
  query {
    mobileImage: file(relativePath: { eq: "foo.png" }) {
      childImageSharp {
        fluid(maxWidth: 600, quality: 100) {
          ...GatsbyImageSharpFluid
        }
      }
    }
    desktopImage: file(
      relativePath: { eq: "foo.png" }
    ) {
      childImageSharp {
        fluid(maxWidth: 800, quality: 100) {
          ...GatsbyImageSharpFluid
        }
      }
    }
  }
```

And I want to note that this is not an issue with `gatsby-image` itself, but an issue with how you are specifying the ``fluid(maxWidth` that is causing duplicate images with same dimensions.

Basically, don't just put any value in `fluid(maxWidth:`. Check the size of your image and then make sure to add valid values to `fluid(maxWidth:` so that the images rendered really have different sizes. Hope that makes sense and solves your problem.

**Comment by visualfanatic** [comment]
@kimbaudi How about giving an option to provide keys the same `media` property is provided?
```js
  const sources = [
    {
      ...post.mobileImage.fluid,
      key: 'mobileImage',
    },
    {
      ...post.xlImage.fluid,
      media: `(min-width: ${theme.breakpoints.xl})`,
      key: 'xlImage',
    },
  ];
```
